### DISCUSSION TITLE
LSP mode format + organize import?

### DISCUSSION THREAD

**Discussion by rickyson96** [radian-software/apheleia#205] [Q&A] [answered]
Hi!

I'm currently following the method suggested by https://github.com/radian-software/apheleia/discussions/120, for enabling `lsp-mode` formatter using `apheleia`

Using the same idea, I'm writing these for organize import
```el
(cl-defun apheleia-lsp-organize-import-formatter (&key buffer scratch callback stdin &allow-other-keys)
      (with-current-buffer buffer
	(when (lsp-feature? "textDocument/codeAction")
	  (if-let ((action (->> (lsp-get-or-calculate-code-actions "source.organizeImports")
				(-filter (-lambda ((&CodeAction :kind?))
					     (and kind? (s-prefix? "source.organizeImports" kind?))))
				lsp--select-action)))
	      (with-current-buffer scratch
		(lsp-execute-code-action action)))))
      (funcall callback))
```

Now, when I use them separately, both performs their own job successfully. But, when I tried to combine them, only one of them runs.
Either `(lsp-format lsp-organize)` or `(lsp-organize lsp-format)` both only runs the `lsp-organize` part.

Is there anything I miss?
Thanks!

**Comment by raxod502** [comment]
Is it possible that `lsp-execute-code-action` is switching back to the original buffer and modifying it, rather than applying to the scratch buffer as in the implementation from https://github.com/radian-software/apheleia/discussions/120? In that case, the original buffer being modified might overwrite the planned changes Apheleia was going to apply, since in the case that the buffer is modified before formatters finish running, the run is abandoned to avoid accidentally reverting recent changes after a save.

Looking at https://github.com/emacs-lsp/lsp-mode/blob/c9db552dcef80c481ecca6bf53c86a12f2a9cd55/lsp-mode.el#L4260, this seems like it might be the case.

**Reply by rickyson96** [reply, marked as answer]
Hey, sorry for just responding to this.
I decided to use another method previously since its available, and currently I'm stuck in the situation where I need this and decided to dig deeper.

Thanks for your pointer, I have working solution right now! Here's the snippet.

```emacs-lisp
(cl-defun apheleia-lsp-organize-import-formatter (&key buffer scratch callback stdin &allow-other-keys)
  (with-current-buffer buffer
	(if-let ((_ (lsp-feature? "textDocument/codeAction"))
			 (action (->> (lsp-get-or-calculate-code-actions "source.organizeImports")
						  lsp--select-action)))
		(-let* ((workspace-edit (cadr action))
				((&WorkspaceEdit :document-changes?) workspace-edit)
				(document-changes (seq-reverse document-changes?)))
		  (lsp--check-document-changes-version document-changes)
		  (->> document-changes
			   (seq-do (lambda (change)
						 (with-current-buffer scratch
						   (lsp--apply-text-edits (lsp:text-document-edit-edits change) 'code-action)))))))
	(funcall callback)))

(add-to-list 'apheleia-formatters '(lsp-organize . apheleia-lsp-organize-import-formatter))
(setf (alist-get 'typescript-ts-mode apheleia-mode-alist) '(lsp-organize eslint))
```
